/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "velox/experimental/cudf/CudfConfig.h"
#include "velox/experimental/cudf/exec/CudfAggregationNode.h"
#include "velox/experimental/cudf/exec/CudfConversion.h"
#include "velox/experimental/cudf/exec/CudfPlanRewriter.h"
#include "velox/experimental/cudf/exec/DecimalAggregationState.h"
#include "velox/experimental/cudf/exec/ToCudf.h"
#include "velox/experimental/cudf/exec/VeloxCudfInterop.h"

#include "velox/common/base/tests/GTestUtils.h"
#include "velox/common/file/FileSystems.h"
#include "velox/exec/PlanNodeStats.h"
#include "velox/exec/tests/utils/AssertQueryBuilder.h"
#include "velox/exec/tests/utils/OperatorTestBase.h"
#include "velox/exec/tests/utils/PlanBuilder.h"
#include "velox/functions/prestosql/aggregates/RegisterAggregateFunctions.h"
#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
#include "velox/parse/TypeResolver.h"
#include "velox/type/DecimalUtil.h"

#include <cudf/column/column_factories.hpp>
#include <cudf/concatenate.hpp>
#include <cudf/copying.hpp>
#include <cudf/null_mask.hpp>
#include <cudf/strings/strings_column_view.hpp>
#include <cudf/utilities/default_stream.hpp>

#include <cuda_runtime_api.h>
#include <rmm/mr/per_device_resource.hpp>

#include <cstdlib>
#include <limits>
#include <optional>
#include <type_traits>
#include <typeinfo>

namespace facebook::velox::cudf_velox {
namespace {

// Tests use cuDF defaults, so do not include the default-poisoning header.
rmm::device_async_resource_ref get_temp_mr() {
  return rmm::mr::get_current_device_resource_ref();
}

int64_t computeAvgRaw(const std::vector<int64_t>& values) {
  int128_t sum = 0;
  for (auto value : values) {
    sum += value;
  }
  int128_t avg = 0;
  facebook::velox::DecimalUtil::computeAverage(avg, sum, values.size(), 0);
  return static_cast<int64_t>(avg);
}

constexpr int kBitsPerWord = 8 * sizeof(cudf::bitmask_type);

std::pair<rmm::device_buffer, cudf::size_type> makeNullMask(
    const std::vector<bool>& valid,
    rmm::cuda_stream_view stream) {
  auto numBits = static_cast<cudf::size_type>(valid.size());
  if (numBits == 0) {
    return {rmm::device_buffer{}, 0};
  }
  auto maskBytes = cudf::bitmask_allocation_size_bytes(numBits);
  auto numWords = maskBytes / sizeof(cudf::bitmask_type);
  std::vector<cudf::bitmask_type> host(numWords, 0);
  cudf::size_type nullCount = 0;
  for (cudf::size_type i = 0; i < numBits; ++i) {
    if (valid[i]) {
      auto word = i / kBitsPerWord;
      auto bit = i % kBitsPerWord;
      host[word] |= (cudf::bitmask_type{1} << bit);
    } else {
      ++nullCount;
    }
  }
  rmm::device_buffer mask(maskBytes, stream);
  if (!host.empty()) {
    auto status = cudaMemcpyAsync(
        mask.data(),
        host.data(),
        host.size() * sizeof(cudf::bitmask_type),
        cudaMemcpyHostToDevice,
        stream.value());
    VELOX_CHECK_EQ(0, static_cast<int>(status));
    stream.synchronize();
  }
  return {std::move(mask), nullCount};
}

class ScopedEnvVar {
 public:
  ScopedEnvVar(const char* key, const char* value) : key_(key) {
    const char* existing = std::getenv(key);
    if (existing) {
      oldValue_ = std::string(existing);
    }
    if (value) {
      setenv(key, value, 1);
    } else {
      unsetenv(key);
    }
  }

  ~ScopedEnvVar() {
    if (oldValue_) {
      setenv(key_.c_str(), oldValue_->c_str(), 1);
    } else {
      unsetenv(key_.c_str());
    }
  }

 private:
  std::string key_;
  std::optional<std::string> oldValue_;
};

template <typename T>
std::unique_ptr<cudf::column> makeFixedWidthColumn(
    cudf::data_type type,
    const std::vector<T>& values,
    const std::vector<bool>* valid,
    rmm::cuda_stream_view stream) {
  auto col = cudf::make_fixed_width_column(
      type,
      static_cast<cudf::size_type>(values.size()),
      cudf::mask_state::UNALLOCATED,
      stream);
  if (!values.empty()) {
    auto status = cudaMemcpyAsync(
        col->mutable_view().data<T>(),
        values.data(),
        values.size() * sizeof(T),
        cudaMemcpyHostToDevice,
        stream.value());
    VELOX_CHECK_EQ(0, static_cast<int>(status));
    stream.synchronize();
  }
  if (valid) {
    auto [mask, nullCount] = makeNullMask(*valid, stream);
    col->set_null_mask(std::move(mask), nullCount);
  }
  return col;
}

template <typename T>
std::unique_ptr<cudf::column> makeDecimalColumn(
    const std::vector<T>& values,
    int32_t scale,
    const std::vector<bool>* valid,
    rmm::cuda_stream_view stream) {
  cudf::type_id typeId = std::is_same_v<T, int64_t> ? cudf::type_id::DECIMAL64
                                                    : cudf::type_id::DECIMAL128;
  cudf::data_type type{typeId, -scale};
  return makeFixedWidthColumn(type, values, valid, stream);
}

std::unique_ptr<cudf::column> makeInt64Column(
    const std::vector<int64_t>& values,
    const std::vector<bool>* valid,
    rmm::cuda_stream_view stream) {
  return makeFixedWidthColumn(
      cudf::data_type{cudf::type_id::INT64}, values, valid, stream);
}

template <typename T>
std::vector<T> copyColumnData(
    const cudf::column_view& view,
    rmm::cuda_stream_view stream) {
  std::vector<T> host(view.size());
  if (view.size() == 0) {
    return host;
  }
  auto status = cudaMemcpyAsync(
      host.data(),
      view.data<T>(),
      view.size() * sizeof(T),
      cudaMemcpyDeviceToHost,
      stream.value());
  VELOX_CHECK_EQ(0, static_cast<int>(status));
  stream.synchronize();
  return host;
}

std::vector<cudf::bitmask_type> copyNullMask(
    const cudf::column_view& view,
    rmm::cuda_stream_view stream) {
  auto numWords = cudf::num_bitmask_words(view.size());
  std::vector<cudf::bitmask_type> host(numWords, 0);
  if (!view.nullable() || numWords == 0) {
    return host;
  }
  auto status = cudaMemcpyAsync(
      host.data(),
      view.null_mask(),
      host.size() * sizeof(cudf::bitmask_type),
      cudaMemcpyDeviceToHost,
      stream.value());
  VELOX_CHECK_EQ(0, static_cast<int>(status));
  stream.synchronize();
  return host;
}

bool isValidAt(const std::vector<cudf::bitmask_type>& mask, size_t idx) {
  if (mask.empty()) {
    return true;
  }
  auto word = idx / kBitsPerWord;
  auto bit = idx % kBitsPerWord;
  return (mask[word] >> bit) & 1;
}

class CudfDecimalTest : public exec::test::OperatorTestBase {
 protected:
  void SetUp() override {
    exec::test::OperatorTestBase::SetUp();
    filesystems::registerLocalFileSystem();
    parse::registerTypeResolver();
    functions::prestosql::registerAllScalarFunctions();
    aggregate::prestosql::registerAllAggregateFunctions();
    CudfConfig::getInstance().allowCpuFallback = false;
    // Ensure a CUDA device is selected and initialized (RMM asserts otherwise).
    int deviceCount = 0;
    auto status = cudaGetDeviceCount(&deviceCount);
    if (status != cudaSuccess) {
      GTEST_SKIP() << "cudaGetDeviceCount failed: " << static_cast<int>(status)
                   << " (" << cudaGetErrorString(status) << ")";
    }
    if (deviceCount == 0) {
      GTEST_SKIP() << "No CUDA devices visible (check CUDA_VISIBLE_DEVICES)";
    }
    VELOX_CHECK_EQ(0, static_cast<int>(cudaSetDevice(0)));
    VELOX_CHECK_EQ(0, static_cast<int>(cudaFree(nullptr)));
    registerCudf();
  }

  void TearDown() override {
    unregisterCudf();
    CudfConfig::getInstance().allowCpuFallback = false;
    exec::test::OperatorTestBase::TearDown();
  }

  void enableCpuFallback() {
    // The driver adapter captures this setting at registration.
    unregisterCudf();
    CudfConfig::getInstance().allowCpuFallback = true;
    registerCudf();
  }
};

TEST_F(CudfDecimalTest, decimalAvgDecimalInput) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeFlatVector<int64_t>(
          {100, 200, 300, 400}, // 1.00, 2.00, 3.00, 4.00
          DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({}, {"avg(d) AS avg_d"})
                  .planNode();

  auto expected = makeRowVector(
      {"avg_d"}, {makeFlatVector<int64_t>({250}, DECIMAL(12, 2))}); // 2.50

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgDecimalInputRounds) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  // Sum = 1.60, count = 7 => 0.22857..., rounds to 0.23 at scale 2.
  std::vector<int64_t> rawValues = {100, 10, 10, 10, 10, 10, 10};
  auto input = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>(rawValues, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({}, {"avg(d) AS avg_d"})
                  .planNode();

  auto expected = makeRowVector(
      {"avg_d"},
      {makeFlatVector<int64_t>({computeAvgRaw(rawValues)}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgPartialFinalVarbinaryRounds) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  std::vector<int32_t> keys = {1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 3};
  std::vector<int64_t> values = {100, 10, 10, 10, 10, 10, 10, 100, 1, -100, -1};

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>(keys),
          makeFlatVector<int64_t>(values, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"avg(d) AS a"})
                  .finalAggregation()
                  .orderBy({"k"}, false)
                  .planNode();

  std::vector<std::pair<int32_t, std::vector<int64_t>>> groups = {
      {1, {100, 10, 10, 10, 10, 10, 10}},
      {2, {100, 1}},
      {3, {-100, -1}},
  };

  auto expected = makeRowVector(
      {"k", "a"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeFlatVector<int64_t>(
              {computeAvgRaw(groups[0].second),
               computeAvgRaw(groups[1].second),
               computeAvgRaw(groups[2].second)},
              DECIMAL(12, 2)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgIntermediateVarbinaryRounds) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 1, 2, 3}),
          makeFlatVector<int64_t>({100, 10, 100, -100}, DECIMAL(12, 2)),
      });
  auto input2 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 1, 1, 1, 1, 2, 3}),
          makeFlatVector<int64_t>({10, 10, 10, 10, 10, 1, -1}, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input1, input2};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"avg(d) AS a"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .orderBy({"k"}, false)
                  .planNode();

  std::vector<std::pair<int32_t, std::vector<int64_t>>> groups = {
      {1, {100, 10, 10, 10, 10, 10, 10}},
      {2, {100, 1}},
      {3, {-100, -1}},
  };

  auto expected = makeRowVector(
      {"k", "a"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeFlatVector<int64_t>(
              {computeAvgRaw(groups[0].second),
               computeAvgRaw(groups[1].second),
               computeAvgRaw(groups[2].second)},
              DECIMAL(12, 2)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalPartialFinalVarbinaryRounds) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({100, 10, 10}, DECIMAL(12, 2))});
  auto input2 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({10, 10, 10, 10}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input1, input2};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"avg(d) AS a"})
                  .finalAggregation()
                  .planNode();

  std::vector<int64_t> allValues = {100, 10, 10, 10, 10, 10, 10};
  auto expected = makeRowVector(
      {"a"},
      {makeFlatVector<int64_t>({computeAvgRaw(allValues)}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalIntermediateVarbinaryRounds) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({100, 10, 10}, DECIMAL(12, 2))});
  auto input2 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({10, 10, 10, 10}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input1, input2};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"avg(d) AS a"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .planNode();

  std::vector<int64_t> allValues = {100, 10, 10, 10, 10, 10, 10};
  auto expected = makeRowVector(
      {"a"},
      {makeFlatVector<int64_t>({computeAvgRaw(allValues)}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalSingleRounds) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeFlatVector<int64_t>({100, 10, 10, 10, 10, 10, 10}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({}, {"avg(d) AS a"})
                  .planNode();

  std::vector<int64_t> allValues = {100, 10, 10, 10, 10, 10, 10};
  auto expected = makeRowVector(
      {"a"},
      {makeFlatVector<int64_t>({computeAvgRaw(allValues)}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalSingleDecimal64Overflow) {
  // 12 values of 9e17 (DECIMAL(18,0)) sum to 1.08e19, past 2^63. The sum must
  // accumulate in 128 bits or a DECIMAL64 accumulator wraps; avg is 9e17.
  constexpr int64_t kBig = 900'000'000'000'000'000;
  constexpr int kNumRows = 12;
  std::vector<int64_t> values(kNumRows, kBig);

  auto input =
      makeRowVector({"d"}, {makeFlatVector<int64_t>(values, DECIMAL(18, 0))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({}, {"avg(d) AS a"})
                  .planNode();

  auto expected =
      makeRowVector({"a"}, {makeFlatVector<int64_t>({kBig}, DECIMAL(18, 0))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalPartialDecimal64Overflow) {
  constexpr int64_t kBig = 900'000'000'000'000'000;
  constexpr int kNumRows = 12;
  std::vector<int64_t> values(kNumRows, kBig);

  auto input =
      makeRowVector({"d"}, {makeFlatVector<int64_t>(values, DECIMAL(18, 0))});

  auto plan = exec::test::PlanBuilder()
                  .values({input})
                  .partialAggregation({}, {"avg(d) AS a"})
                  .finalAggregation()
                  .planNode();

  auto expected =
      makeRowVector({"a"}, {makeFlatVector<int64_t>({kBig}, DECIMAL(18, 0))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGroupbySingleDecimal64Overflow) {
  // Same overflow within a single group, exercising the groupby raw sum path.
  constexpr int64_t kBig = 900'000'000'000'000'000;
  constexpr int kNumRows = 12;
  std::vector<int32_t> keys(kNumRows, 1);
  std::vector<int64_t> values(kNumRows, kBig);

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>(keys),
          makeFlatVector<int64_t>(values, DECIMAL(18, 0)),
      });

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({"k"}, {"avg(d) AS a"})
                  .planNode();

  auto expected = makeRowVector(
      {"k", "a"},
      {
          makeFlatVector<int32_t>({1}),
          makeFlatVector<int64_t>({kBig}, DECIMAL(18, 0)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalSingleAllNulls) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeNullableFlatVector<int64_t>(
          {std::nullopt, std::nullopt, std::nullopt}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({}, {"avg(d) AS a"})
                  .planNode();

  auto expected = makeRowVector(
      {"a"}, {makeNullableFlatVector<int64_t>({std::nullopt}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalPartialFinalVarbinaryAllNulls) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeNullableFlatVector<int64_t>(
          {std::nullopt, std::nullopt, std::nullopt}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"avg(d) AS a"})
                  .finalAggregation()
                  .planNode();

  auto expected = makeRowVector(
      {"a"}, {makeNullableFlatVector<int64_t>({std::nullopt}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgGlobalIntermediateVarbinaryAllNulls) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeNullableFlatVector<int64_t>(
          {std::nullopt, std::nullopt, std::nullopt}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"avg(d) AS a"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .planNode();

  auto expected = makeRowVector(
      {"a"}, {makeNullableFlatVector<int64_t>({std::nullopt}, DECIMAL(12, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgPartialFinalVarbinaryNullGroup) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 1, 2, 2, 3, 3}),
          makeNullableFlatVector<int64_t>(
              {100, 200, std::nullopt, std::nullopt, 400, std::nullopt},
              DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"avg(d) AS a"})
                  .finalAggregation()
                  .orderBy({"k"}, false)
                  .planNode();

  auto expected = makeRowVector(
      {"k", "a"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int64_t>(
              {150, std::nullopt, 400}, DECIMAL(12, 2)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalAvgIntermediateVarbinaryNullGroup) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int64_t>(
              {100, std::nullopt, 400}, DECIMAL(12, 2)),
      });
  auto input2 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int64_t>(
              {200, std::nullopt, std::nullopt}, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input1, input2};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"avg(d) AS a"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .orderBy({"k"}, false)
                  .planNode();

  auto expected = makeRowVector(
      {"k", "a"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int64_t>(
              {150, std::nullopt, 400}, DECIMAL(12, 2)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, streamingDecimalSumAndAvg) {
  auto& config = CudfConfig::getInstance();
  const auto savedEnabled = config.streamingGroupbyApiEnabled;
  config.streamingGroupbyApiEnabled = true;
  SCOPE_EXIT {
    config.streamingGroupbyApiEnabled = savedEnabled;
  };

  auto first = makeRowVector({
      makeFlatVector<int32_t>({0, 1}),
      makeFlatVector<int64_t>({999'999'999'999'999'999, -200}, DECIMAL(18, 2)),
      makeFlatVector<int128_t>({100, -200}, DECIMAL(20, 2)),
  });
  auto later = makeRowVector({
      makeFlatVector<int32_t>({0, 1, 2, 3}),
      makeNullableFlatVector<int64_t>(
          {999'999'999'999'999'999, -200, std::nullopt, 0}, DECIMAL(18, 2)),
      makeNullableFlatVector<int128_t>(
          {100, -200, std::nullopt, 0}, DECIMAL(20, 2)),
  });
  std::vector<RowVectorPtr> batches{first};
  batches.insert(batches.end(), 11, later);
  for (bool compact : {false, true}) {
    for (bool includeAvg : {false, true}) {
      SCOPED_TRACE(fmt::format("compact={}, avg={}", compact, includeAvg));
      std::vector<std::string> aggregates{"sum(c1)", "sum(c2)"};
      std::vector<VectorPtr> expectedColumns{
          makeFlatVector<int32_t>({0, 1, 2, 3}),
          makeNullableFlatVector<int128_t>(
              {int128_t{999'999'999'999'999'999} * 12, -2400, std::nullopt, 0},
              DECIMAL(38, 2)),
          makeNullableFlatVector<int128_t>(
              {1200, -2400, std::nullopt, 0}, DECIMAL(38, 2))};
      if (includeAvg) {
        aggregates.push_back("avg(c1)");
        aggregates.push_back("avg(c2)");
        expectedColumns.push_back(
            makeNullableFlatVector<int64_t>(
                {999'999'999'999'999'999, -200, std::nullopt, 0},
                DECIMAL(18, 2)));
        expectedColumns.push_back(
            makeNullableFlatVector<int128_t>(
                {100, -200, std::nullopt, 0}, DECIMAL(20, 2)));
      }
      auto plan = exec::test::PlanBuilder()
                      .values(batches)
                      .partialAggregation({"c0"}, aggregates)
                      .finalAggregation()
                      .planNode();
      if (compact) {
        plan = CudfPlanRewriter::rewrite(plan);
        auto node = dynamic_cast<const CudfAggregationNode*>(plan.get());
        ASSERT_NE(node, nullptr);
        EXPECT_TRUE(node->usesCompactDecimalSum(0));
      }
      auto task =
          exec::test::AssertQueryBuilder(plan)
              .maxDrivers(1)
              .config(CudfFromVelox::kGpuBatchSizeRows, 1)
              .config(core::QueryConfig::kMaxPartialAggregationMemory, 1)
              .assertResults(makeRowVector(expectedColumns));
      const auto stats = exec::toPlanStats(task->taskStats());
      const auto& customStats = stats.at(plan->id()).customStats;
      EXPECT_GT(customStats.at("streamingGroupbyApiUsed").sum, 0);
      EXPECT_GT(customStats.at("streamingGroupbyApiRebuilds").sum, 0);
    }
  }
}

TEST_F(CudfDecimalTest, streamingDecimalAvgUnequalCounts) {
  auto& config = CudfConfig::getInstance();
  const auto savedEnabled = config.streamingGroupbyApiEnabled;
  config.streamingGroupbyApiEnabled = true;
  SCOPE_EXIT {
    config.streamingGroupbyApiEnabled = savedEnabled;
  };
  auto keys = makeFlatVector<int32_t>({0, 0, 1, 1, 2, 2});
  auto first = makeRowVector({
      keys,
      makeNullableFlatVector<int64_t>(
          {101, 102, -101, -102, std::nullopt, std::nullopt}, DECIMAL(12, 2)),
      makeNullableFlatVector<int128_t>(
          {101, 102, -101, -102, std::nullopt, std::nullopt}, DECIMAL(20, 2)),
  });
  auto second = makeRowVector({
      keys,
      makeNullableFlatVector<int64_t>(
          {104, std::nullopt, -104, std::nullopt, std::nullopt, std::nullopt},
          DECIMAL(12, 2)),
      makeNullableFlatVector<int128_t>(
          {104, std::nullopt, -104, std::nullopt, std::nullopt, std::nullopt},
          DECIMAL(20, 2)),
  });
  auto expected = makeRowVector({
      makeFlatVector<int32_t>({0, 1, 2}),
      makeNullableFlatVector<int64_t>(
          {102, -102, std::nullopt}, DECIMAL(12, 2)),
      makeNullableFlatVector<int128_t>(
          {102, -102, std::nullopt}, DECIMAL(20, 2)),
  });
  auto plan = exec::test::PlanBuilder()
                  .values({first, second})
                  .partialAggregation({"c0"}, {"avg(c1)", "avg(c2)"})
                  .finalAggregation()
                  .planNode();
  auto task = exec::test::AssertQueryBuilder(plan)
                  .maxDrivers(1)
                  .config(CudfFromVelox::kGpuBatchSizeRows, 6)
                  .config(core::QueryConfig::kMaxPartialAggregationMemory, 1)
                  .assertResults(expected);
  const auto stats = exec::toPlanStats(task->taskStats());
  EXPECT_GT(
      stats.at(plan->id()).customStats.at("streamingGroupbyApiUsed").sum, 0);
}

TEST_F(CudfDecimalTest, directFinalizationDecimalSumAndAvg) {
  auto& config = CudfConfig::getInstance();
  const auto savedEnabled = config.streamingGroupbyApiEnabled;
  config.streamingGroupbyApiEnabled = false;
  SCOPE_EXIT {
    config.streamingGroupbyApiEnabled = savedEnabled;
  };
  auto input = makeRowVector({
      makeFlatVector<int32_t>({1, 1, 2}),
      makeNullableFlatVector<int64_t>({100, 300, std::nullopt}, DECIMAL(12, 2)),
      makeNullableFlatVector<int128_t>(
          {100, 300, std::nullopt}, DECIMAL(20, 2)),
  });
  auto expected = makeRowVector({
      makeFlatVector<int32_t>({1, 2}),
      makeNullableFlatVector<int128_t>({800, std::nullopt}, DECIMAL(38, 2)),
      makeNullableFlatVector<int64_t>({200, std::nullopt}, DECIMAL(12, 2)),
      makeNullableFlatVector<int128_t>({800, std::nullopt}, DECIMAL(38, 2)),
      makeNullableFlatVector<int128_t>({200, std::nullopt}, DECIMAL(20, 2)),
  });
  const std::vector<std::string> aggregates{
      "sum(c1)", "avg(c1)", "sum(c2)", "avg(c2)"};
  for (bool single : {true, false}) {
    for (bool compact : {false, true}) {
      if (single && compact) {
        continue;
      }
      SCOPED_TRACE(fmt::format("single={}, compact={}", single, compact));
      auto builder = exec::test::PlanBuilder().values({input, input});
      if (single) {
        builder.singleAggregation({"c0"}, aggregates);
      } else {
        builder.partialAggregation({"c0"}, aggregates).finalAggregation();
      }
      auto plan = builder.planNode();
      if (compact) {
        plan = CudfPlanRewriter::rewrite(plan);
      }
      auto task =
          exec::test::AssertQueryBuilder(plan)
              .maxDrivers(1)
              .config(CudfFromVelox::kGpuBatchSizeRows, "3")
              .config(core::QueryConfig::kMaxPartialAggregationMemory, 1)
              .assertResults(expected);
      const auto stats = exec::toPlanStats(task->taskStats());
      EXPECT_GT(
          stats.at(plan->id()).customStats.at("directGroupbyFinalization").sum,
          0);
    }
  }
}

TEST_F(CudfDecimalTest, decimalSumPartialFinalVarbinary) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 1, 2, 2, 2}),
          makeFlatVector<int64_t>(
              {12345, -2500, 10000, 200, -300}, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input};
  createDuckDbTable(vectors);

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"sum(d) AS s"})
                  .finalAggregation()
                  .planNode();

  facebook::velox::exec::test::AssertQueryBuilder(plan, duckDbQueryRunner_)
      .assertResults("SELECT k, sum(d) AS s FROM tmp GROUP BY k");
}

TEST_F(CudfDecimalTest, decimalPartialSumVarbinaryToVeloxRoundTrip) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({100, 200, 300}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"sum(d) AS s"})
                  .planNode();

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  VELOX_CHECK_NOT_NULL(result);
  ASSERT_GT(result->size(), 0);
  ASSERT_EQ(result->childAt(0)->type()->kind(), TypeKind::VARBINARY);
}

TEST_F(CudfDecimalTest, decimalSumPartialFinalEmptyInput) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeFlatVector<int64_t>({100, 200, 300}, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input};
  createDuckDbTable(vectors);

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .filter("k < 0")
                  .partialAggregation({"k"}, {"sum(d) AS s"})
                  .finalAggregation()
                  .planNode();

  facebook::velox::exec::test::AssertQueryBuilder(plan, duckDbQueryRunner_)
      .assertResults("SELECT k, sum(d) AS s FROM tmp WHERE k < 0 GROUP BY k");
}

TEST_F(CudfDecimalTest, decimalSumIntermediateVarbinary) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 1, 2}),
          makeFlatVector<int64_t>({12345, -2500, 10000}, DECIMAL(12, 2)),
      });
  auto input2 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({2, 3}),
          makeFlatVector<int64_t>({200, -300}, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input1, input2};
  createDuckDbTable(vectors);

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"sum(d) AS s"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .planNode();

  facebook::velox::exec::test::AssertQueryBuilder(plan, duckDbQueryRunner_)
      .assertResults("SELECT k, sum(d) AS s FROM tmp GROUP BY k");
}

TEST_F(CudfDecimalTest, decimalSumGlobalPartialFinalVarbinary) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({12345, -2500, 10000}, DECIMAL(12, 2))});
  auto input2 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({200, -300}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input1, input2};
  createDuckDbTable(vectors);

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"sum(d) AS s"})
                  .finalAggregation()
                  .planNode();

  facebook::velox::exec::test::AssertQueryBuilder(plan, duckDbQueryRunner_)
      .assertResults("SELECT sum(d) AS s FROM tmp");
}

TEST_F(CudfDecimalTest, decimalSumGlobalIntermediateVarbinary) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({12345, -2500, 10000}, DECIMAL(12, 2))});
  auto input2 = makeRowVector(
      {"d"}, {makeFlatVector<int64_t>({200, -300}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input1, input2};
  createDuckDbTable(vectors);

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"sum(d) AS s"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .planNode();

  facebook::velox::exec::test::AssertQueryBuilder(plan, duckDbQueryRunner_)
      .assertResults("SELECT sum(d) AS s FROM tmp");
}

TEST_F(CudfDecimalTest, decimalSumGlobalSingle) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeFlatVector<int64_t>(
          {12345, -2500, 10000, 200, -300}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};
  createDuckDbTable(vectors);

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({}, {"sum(d) AS s"})
                  .planNode();

  facebook::velox::exec::test::AssertQueryBuilder(plan, duckDbQueryRunner_)
      .assertResults("SELECT sum(d) AS s FROM tmp");
}

TEST_F(CudfDecimalTest, decimalSumGroupbySingleDecimal64Overflow) {
  // One group of 12 values of 9e17 (DECIMAL(18,0)) sums to 1.08e19, past 2^63.
  // sum(decimal(18,0)) -> decimal(38,0), computed in 128 bits, no wrap.
  constexpr int64_t kBig = 900'000'000'000'000'000;
  constexpr int kNumRows = 12;
  std::vector<int32_t> keys(kNumRows, 1);
  std::vector<int64_t> values(kNumRows, kBig);

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>(keys),
          makeFlatVector<int64_t>(values, DECIMAL(18, 0)),
      });

  std::vector<RowVectorPtr> vectors = {input};

  const int128_t expectedSum = static_cast<int128_t>(kBig) * kNumRows;
  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .singleAggregation({"k"}, {"sum(d) AS s"})
                  .planNode();

  auto expected = makeRowVector(
      {"k", "s"},
      {
          makeFlatVector<int32_t>({1}),
          makeFlatVector<int128_t>({expectedSum}, DECIMAL(38, 0)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalSumGlobalSingleDecimal64Overflow) {
  constexpr int64_t kBig = 900'000'000'000'000'000;
  constexpr int kNumRows = 12;
  std::vector<int64_t> values(kNumRows, kBig);

  auto input =
      makeRowVector({"d"}, {makeFlatVector<int64_t>(values, DECIMAL(18, 0))});

  const int128_t expectedSum = static_cast<int128_t>(kBig) * kNumRows;
  auto plan = exec::test::PlanBuilder()
                  .values({input})
                  .singleAggregation({}, {"sum(d) AS s"})
                  .planNode();

  auto expected = makeRowVector(
      {"s"}, {makeFlatVector<int128_t>({expectedSum}, DECIMAL(38, 0))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);

  auto compactPlan = exec::test::PlanBuilder()
                         .values({input})
                         .partialAggregation({}, {"sum(d) AS s"})
                         .finalAggregation()
                         .planNode();
  exec::test::AssertQueryBuilder(CudfPlanRewriter::rewrite(compactPlan))
      .assertResults(expected);
}

TEST_F(CudfDecimalTest, decimalSumGlobalPartialFinalDecimal64Overflow) {
  // Global SUM whose total overflows DECIMAL64; exercises the partial raw sum
  // (serialized to VARBINARY) and the final merge, both in 128 bits.
  constexpr int64_t kBig = 900'000'000'000'000'000;
  constexpr int kNumRows = 12;
  std::vector<int64_t> values(kNumRows, kBig);

  auto input =
      makeRowVector({"d"}, {makeFlatVector<int64_t>(values, DECIMAL(18, 0))});

  std::vector<RowVectorPtr> vectors = {input};

  const int128_t expectedSum = static_cast<int128_t>(kBig) * kNumRows;
  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"sum(d) AS s"})
                  .finalAggregation()
                  .planNode();

  auto expected = makeRowVector(
      {"s"}, {makeFlatVector<int128_t>({expectedSum}, DECIMAL(38, 0))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalSumPartialFinalVarbinaryNullGroup) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 1, 2, 2, 3, 3}),
          makeNullableFlatVector<int64_t>(
              {100, 200, std::nullopt, std::nullopt, 400, std::nullopt},
              DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"sum(d) AS s"})
                  .finalAggregation()
                  .orderBy({"k"}, false)
                  .planNode();

  auto expected = makeRowVector(
      {"k", "s"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int128_t>(
              {static_cast<int128_t>(300),
               std::nullopt,
               static_cast<int128_t>(400)},
              DECIMAL(38, 2)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalSumIntermediateVarbinaryNullGroup) {
  auto rowType = ROW({
      {"k", INTEGER()},
      {"d", DECIMAL(12, 2)},
  });

  auto input1 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int64_t>(
              {100, std::nullopt, 400}, DECIMAL(12, 2)),
      });
  auto input2 = makeRowVector(
      {"k", "d"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int64_t>(
              {200, std::nullopt, std::nullopt}, DECIMAL(12, 2)),
      });

  std::vector<RowVectorPtr> vectors = {input1, input2};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({"k"}, {"sum(d) AS s"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .orderBy({"k"}, false)
                  .planNode();

  auto expected = makeRowVector(
      {"k", "s"},
      {
          makeFlatVector<int32_t>({1, 2, 3}),
          makeNullableFlatVector<int128_t>(
              {static_cast<int128_t>(300),
               std::nullopt,
               static_cast<int128_t>(400)},
              DECIMAL(38, 2)),
      });

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalSumGlobalPartialFinalVarbinaryAllNulls) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeNullableFlatVector<int64_t>(
          {std::nullopt, std::nullopt, std::nullopt}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"sum(d) AS s"})
                  .finalAggregation()
                  .planNode();

  auto expected = makeRowVector(
      {"s"},
      {makeNullableFlatVector<int128_t>({std::nullopt}, DECIMAL(38, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, decimalSumGlobalIntermediateVarbinaryAllNulls) {
  auto rowType = ROW({
      {"d", DECIMAL(12, 2)},
  });

  auto input = makeRowVector(
      {"d"},
      {makeNullableFlatVector<int64_t>(
          {std::nullopt, std::nullopt, std::nullopt}, DECIMAL(12, 2))});

  std::vector<RowVectorPtr> vectors = {input};

  auto plan = exec::test::PlanBuilder()
                  .values(vectors)
                  .partialAggregation({}, {"sum(d) AS s"})
                  .intermediateAggregation()
                  .finalAggregation()
                  .planNode();

  auto expected = makeRowVector(
      {"s"},
      {makeNullableFlatVector<int128_t>({std::nullopt}, DECIMAL(38, 2))});

  auto result =
      facebook::velox::exec::test::AssertQueryBuilder(plan).copyResults(pool());
  facebook::velox::test::assertEqualVectors(expected, result);
}

TEST_F(CudfDecimalTest, compactSumGroupedMixedAggregates) {
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1, 1, 2, 2, 3, 3}),
       makeNullableFlatVector<int64_t>(
           {100, -100, std::nullopt, std::nullopt, 400, std::nullopt},
           DECIMAL(12, 2))});
  auto expected = makeRowVector(
      {"k", "s", "a"},
      {makeFlatVector<int32_t>({1, 2, 3}),
       makeNullableFlatVector<int128_t>({0, std::nullopt, 400}, DECIMAL(38, 2)),
       makeNullableFlatVector<int64_t>(
           {0, std::nullopt, 400}, DECIMAL(12, 2))});

  for (bool intermediate : {false, true}) {
    SCOPED_TRACE(intermediate);
    auto builder = exec::test::PlanBuilder();
    builder.values({input})
        .partialAggregation({"k"}, {"sum(d) AS s", "avg(d) AS a"})
        .localPartition({"k"});
    if (intermediate) {
      builder.intermediateAggregation();
    }
    auto original = builder.finalAggregation().planNode();
    auto rewritten = CudfPlanRewriter::rewrite(original);
    auto node = rewritten;
    size_t stages = 0;
    while (auto* aggregation =
               dynamic_cast<const CudfAggregationNode*>(node.get())) {
      ++stages;
      EXPECT_TRUE(aggregation->usesCompactDecimalSum(0));
      EXPECT_FALSE(aggregation->usesCompactDecimalSum(1));
      EXPECT_EQ(*aggregation->intermediateType(0), *DECIMAL(38, 2));
      node = node->sources()[0];
      if (auto* partition =
              dynamic_cast<const core::LocalPartitionNode*>(node.get())) {
        EXPECT_EQ(partition->name(), "LocalPartition");
        EXPECT_EQ(*partition->outputType()->childAt(1), *DECIMAL(38, 2));
        EXPECT_EQ(*partition->outputType()->childAt(2), *VARBINARY());
        node = partition->sources()[0];
      }
    }
    EXPECT_EQ(stages, intermediate ? 3 : 2);
    ASSERT_NE(dynamic_cast<const core::ValuesNode*>(node.get()), nullptr);
    EXPECT_EQ(*original->outputType(), *rewritten->outputType());
    exec::test::AssertQueryBuilder(rewritten).maxDrivers(2).assertResults(
        expected);
  }
}

TEST_F(CudfDecimalTest, compactSumProjectedLocalPartition) {
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1, 1, 2, 2, 3, 3}),
       makeNullableFlatVector<int64_t>(
           {100, -100, std::nullopt, std::nullopt, 400, std::nullopt},
           DECIMAL(12, 2))});
  auto expected = makeRowVector(
      {"key", "s", "a", "s2"},
      {makeFlatVector<int32_t>({1, 2, 3}),
       makeNullableFlatVector<int128_t>({0, std::nullopt, 800}, DECIMAL(38, 2)),
       makeNullableFlatVector<int64_t>({0, std::nullopt, 400}, DECIMAL(12, 2)),
       makeNullableFlatVector<int128_t>(
           {0, std::nullopt, 800}, DECIMAL(38, 2))});
  auto original =
      exec::test::PlanBuilder()
          .values({input, input})
          .partialAggregation({"k"}, {"sum(d) AS s", "avg(d) AS a"})
          .project(
              {"a AS avg_state", "s AS sum_state", "k AS key", "s AS sum_copy"})
          .localPartition({"key"})
          .finalAggregation(
              {"key"},
              {"sum(sum_state) AS s",
               "avg(avg_state) AS a",
               "sum(sum_copy) AS s2"},
              {{DECIMAL(12, 2)}, {DECIMAL(12, 2)}, {DECIMAL(12, 2)}})
          .planNode();
  auto rewritten = CudfPlanRewriter::rewrite(original);
  const auto& project = rewritten->sources()[0]->sources()[0];
  ASSERT_NE(dynamic_cast<const core::ProjectNode*>(project.get()), nullptr);
  EXPECT_EQ(project->id(), original->sources()[0]->sources()[0]->id());
  EXPECT_EQ(
      *project->outputType(),
      *ROW(
          {"avg_state", "sum_state", "key", "sum_copy"},
          {VARBINARY(), DECIMAL(38, 2), INTEGER(), DECIMAL(38, 2)}));
  EXPECT_EQ(*rewritten->outputType(), *original->outputType());
  EXPECT_EQ(CudfPlanRewriter::rewrite(rewritten), rewritten);
  exec::test::AssertQueryBuilder(rewritten).maxDrivers(2).assertResults(
      expected);
}

TEST_F(CudfDecimalTest, compactSumGlobalNullZeroAndEmpty) {
  const std::vector<std::vector<std::optional<int64_t>>> cases = {
      {100, -100, std::nullopt},
      {std::nullopt, std::nullopt},
      {},
      {999999999999LL, 999999999999LL}};
  const std::vector<std::optional<int128_t>> sums = {
      0, std::nullopt, std::nullopt, int128_t{1999999999998LL}};
  for (size_t i = 0; i < cases.size(); ++i) {
    SCOPED_TRACE(i);
    auto input = makeRowVector(
        {"d"}, {makeNullableFlatVector<int64_t>(cases[i], DECIMAL(12, 2))});
    auto expected = makeRowVector(
        {"s"}, {makeNullableFlatVector<int128_t>({sums[i]}, DECIMAL(38, 2))});
    for (bool intermediate : {false, true}) {
      SCOPED_TRACE(intermediate);
      auto builder = exec::test::PlanBuilder();
      builder.values({input}).partialAggregation({}, {"sum(d) AS s"});
      if (intermediate) {
        builder.intermediateAggregation();
      }
      auto rewritten =
          CudfPlanRewriter::rewrite(builder.finalAggregation().planNode());
      ASSERT_NE(
          dynamic_cast<const CudfAggregationNode*>(rewritten.get()), nullptr);
      EXPECT_EQ(
          *rewritten->sources()[0]->outputType()->childAt(0), *DECIMAL(38, 2));
      exec::test::AssertQueryBuilder(rewritten).assertResults(expected);
    }
  }
}

TEST_F(CudfDecimalTest, compactSumGroupedEmpty) {
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1}),
       makeFlatVector<int64_t>({100}, DECIMAL(12, 2))});
  auto original = exec::test::PlanBuilder()
                      .values({input})
                      .filter("k < 0")
                      .partialAggregation({"k"}, {"sum(d) AS s"})
                      .intermediateAggregation()
                      .finalAggregation()
                      .planNode();
  auto rewritten = CudfPlanRewriter::rewrite(original);
  auto expected = makeRowVector(
      {"k", "s"},
      {makeFlatVector<int32_t>({}),
       makeFlatVector<int128_t>({}, DECIMAL(38, 2))});
  exec::test::AssertQueryBuilder(rewritten).assertResults(expected);
}

TEST_F(CudfDecimalTest, compactSumSurroundingNodePlans) {
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1, 1, 2}),
       makeFlatVector<int64_t>({100, 200, 300}, DECIMAL(12, 2))});
  auto ids = std::make_shared<core::PlanNodeIdGenerator>();
  auto makeAggregation = [&](bool compact) {
    auto builder = exec::test::PlanBuilder(ids).values({input});
    if (compact) {
      builder.partialAggregation({"k"}, {"sum(d) AS s"}).finalAggregation();
    } else {
      builder.singleAggregation({"k"}, {"sum(d) AS s"});
    }
    return builder.planNode();
  };
  auto checkPreserved = [&](auto&& self,
                            const core::PlanNodePtr& original,
                            const core::PlanNodePtr& rewritten) -> void {
    EXPECT_EQ(original->id(), rewritten->id());
    EXPECT_EQ(*original->outputType(), *rewritten->outputType());
    if (auto* aggregation =
            dynamic_cast<const core::AggregationNode*>(original.get());
        aggregation &&
        aggregation->step() == core::AggregationNode::Step::kFinal) {
      auto* compact = dynamic_cast<const CudfAggregationNode*>(rewritten.get());
      ASSERT_NE(compact, nullptr);
      EXPECT_TRUE(compact->usesCompactDecimalSum(0));
      EXPECT_EQ(
          *compact->sources()[0]->outputType()->childAt(1), *DECIMAL(38, 2));
      return;
    }
    EXPECT_EQ(typeid(*original), typeid(*rewritten));
    auto originalOptions = original->serialize();
    auto rewrittenOptions = rewritten->serialize();
    originalOptions.erase("sources");
    rewrittenOptions.erase("sources");
    EXPECT_EQ(originalOptions, rewrittenOptions);
    ASSERT_EQ(original->sources().size(), rewritten->sources().size());
    for (size_t i = 0; i < original->sources().size(); ++i) {
      self(self, original->sources()[i], rewritten->sources()[i]);
    }
  };

  for (bool compactLeft : {false, true}) {
    for (bool compactRight : {false, true}) {
      SCOPED_TRACE(fmt::format("left={}, right={}", compactLeft, compactRight));
      auto left = makeAggregation(compactLeft);
      auto right = exec::test::PlanBuilder(makeAggregation(compactRight), ids)
                       .project({"k AS rk", "s AS rs"})
                       .planNode();
      std::vector<core::PlanNodePtr> plans;
      for (bool partial : {false, true}) {
        plans.push_back(
            exec::test::PlanBuilder(left, ids)
                .topN({"s DESC NULLS FIRST", "k ASC"}, 2, partial)
                .planNode());
        plans.push_back(
            exec::test::PlanBuilder(left, ids)
                .orderBy({"s DESC NULLS FIRST", "k ASC"}, partial)
                .limit(1, 2, partial)
                .planNode());
      }
      plans.push_back(
          exec::test::PlanBuilder(left, ids)
              .hashJoin(
                  {"k"},
                  {"rk"},
                  right,
                  "k > 0",
                  {"rs", "k", "s"},
                  core::JoinType::kLeft)
              .planNode());
      plans.push_back(
          exec::test::PlanBuilder(left, ids)
              .orderBy({"k"}, false)
              .mergeJoin(
                  {"k"},
                  {"rk"},
                  exec::test::PlanBuilder(right, ids)
                      .orderBy({"rk"}, false)
                      .planNode(),
                  "k > 0",
                  {"rs", "k", "s"},
                  core::JoinType::kLeft)
              .planNode());
      plans.push_back(
          exec::test::PlanBuilder(left, ids)
              .nestedLoopJoin(
                  right, "k < rk", {"rs", "k", "s"}, core::JoinType::kLeft)
              .planNode());
      plans.push_back(
          exec::test::PlanBuilder(ids)
              .localMerge(
                  {"s DESC NULLS FIRST", "k ASC"},
                  {exec::test::PlanBuilder(left, ids)
                       .orderBy({"s DESC NULLS FIRST", "k ASC"}, true)
                       .planNode(),
                   exec::test::PlanBuilder(makeAggregation(compactRight), ids)
                       .orderBy({"s DESC NULLS FIRST", "k ASC"}, true)
                       .planNode()})
              .planNode());
      for (const auto& plan : plans) {
        SCOPED_TRACE(plan->toString(true, true));
        const auto original = plan->serialize();
        auto rewritten = CudfPlanRewriter::rewrite(plan);
        checkPreserved(checkPreserved, plan, rewritten);
        EXPECT_EQ(CudfPlanRewriter::rewrite(rewritten), rewritten);
        EXPECT_EQ(plan->serialize(), original);
        if (!compactLeft && !compactRight) {
          EXPECT_EQ(rewritten, plan);
        }
      }
    }
  }
}

TEST_F(CudfDecimalTest, compactSumSurroundingNodeResults) {
  enableCpuFallback();
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1, 1, 2, 2, 3, 4}),
       makeNullableFlatVector<int64_t>(
           {100, -100, 300, 400, std::nullopt, -200}, DECIMAL(12, 2))});
  auto ids = std::make_shared<core::PlanNodeIdGenerator>();
  auto aggregation = exec::test::PlanBuilder(ids)
                         .values({input})
                         .partialAggregation({"k"}, {"sum(d) AS s"})
                         .finalAggregation()
                         .planNode();
  auto right =
      exec::test::PlanBuilder(ids)
          .values({makeRowVector({"rk"}, {makeFlatVector<int32_t>({1, 2, 4})})})
          .planNode();
  const std::vector<core::PlanNodePtr> plans = {
      exec::test::PlanBuilder(aggregation, ids)
          .topN({"s DESC NULLS FIRST", "k ASC"}, 3, false)
          .planNode(),
      exec::test::PlanBuilder(aggregation, ids)
          .orderBy({"s DESC NULLS FIRST", "k ASC"}, false)
          .limit(1, 2, false)
          .planNode(),
      exec::test::PlanBuilder(aggregation, ids)
          .hashJoin({"k"}, {"rk"}, right, "", {"k", "s"})
          .planNode(),
      exec::test::PlanBuilder(aggregation, ids)
          .nestedLoopJoin(right, "k < rk", {"k", "s", "rk"})
          .planNode()};
  for (size_t i = 0; i < plans.size(); ++i) {
    const auto& plan = plans[i];
    SCOPED_TRACE(plan->toString(true, true));
    auto expected = exec::test::AssertQueryBuilder(plan)
                        .config(CudfConfig::kCudfEnabled, false)
                        .maxDrivers(1)
                        .copyResults(pool());
    auto query =
        exec::test::AssertQueryBuilder(CudfPlanRewriter::rewrite(plan));
    query.config(CudfConfig::kCudfEnabled, true).maxDrivers(1);
    if (i < 2) {
      facebook::velox::test::assertEqualVectors(
          expected, query.copyResults(pool()));
    } else {
      query.assertResults(expected);
    }
  }
}

TEST_F(CudfDecimalTest, compactSumRejectsSurroundingNodesOnPartialState) {
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1}),
       makeFlatVector<int64_t>({100}, DECIMAL(12, 2))});
  auto ids = std::make_shared<core::PlanNodeIdGenerator>();
  auto partial = exec::test::PlanBuilder(ids)
                     .values({input})
                     .partialAggregation({"k"}, {"sum(d) AS s"})
                     .planNode();
  auto right =
      exec::test::PlanBuilder(ids)
          .values({makeRowVector({"rk"}, {makeFlatVector<int32_t>({1})})})
          .planNode();
  const std::vector<core::PlanNodePtr> plans = {
      exec::test::PlanBuilder(partial, ids).topN({"k"}, 1, false).planNode(),
      exec::test::PlanBuilder(partial, ids).limit(0, 1, false).planNode(),
      exec::test::PlanBuilder(partial, ids)
          .hashJoin({"k"}, {"rk"}, right, "", {"k", "s"})
          .planNode(),
      exec::test::PlanBuilder(right, ids)
          .hashJoin({"rk"}, {"k"}, partial, "", {"k", "s"})
          .planNode()};
  for (const auto& plan : plans) {
    SCOPED_TRACE(plan->toString(true, true));
    VELOX_ASSERT_THROW(CudfPlanRewriter::rewrite(plan), "compact");
    auto final =
        exec::test::PlanBuilder(plan, ids)
            .finalAggregation({"k"}, {"sum(s) AS s"}, {{DECIMAL(12, 2)}})
            .planNode();
    VELOX_ASSERT_THROW(
        CudfPlanRewriter::rewrite(final),
        "Unsupported compact aggregation state source");
  }
}

TEST_F(CudfDecimalTest, compactSumUcxFragmentSchemasAndSerde) {
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1}),
       makeFlatVector<int64_t>({100}, DECIMAL(12, 2))});
  auto producer = exec::test::PlanBuilder()
                      .values({input})
                      .partialAggregation({"k"}, {"sum(d) AS s", "avg(d) AS a"})
                      .project({"a AS avg_state", "k", "s AS sum_state"})
                      .partitionedOutput(
                          {"k"},
                          2,
                          false,
                          {},
                          "Presto",
                          std::string{core::TransportKind::kUcx})
                      .planNode();
  auto consumer = exec::test::PlanBuilder()
                      .exchange(
                          producer->outputType(),
                          "Presto",
                          std::string{core::TransportKind::kUcx})
                      .project({"sum_state AS s", "k", "avg_state AS a"})
                      .finalAggregation(
                          {"k"},
                          {"sum(s) AS s", "avg(a) AS a"},
                          {{DECIMAL(12, 2)}, {DECIMAL(12, 2)}})
                      .planNode();
  auto gpuProducer = CudfPlanRewriter::rewrite(producer);
  auto gpuConsumer = CudfPlanRewriter::rewrite(consumer);
  auto* output =
      dynamic_cast<const core::PartitionedOutputNode*>(gpuProducer.get());
  auto* exchange = dynamic_cast<const core::ExchangeNode*>(
      gpuConsumer->sources()[0]->sources()[0].get());
  ASSERT_NE(output, nullptr);
  ASSERT_NE(exchange, nullptr);
  EXPECT_EQ(output->name(), producer->name());
  EXPECT_EQ(exchange->name(), consumer->sources()[0]->sources()[0]->name());
  EXPECT_EQ(output->transportKind(), core::TransportKind::kUcx);
  EXPECT_EQ(exchange->transportKind(), core::TransportKind::kUcx);
  EXPECT_EQ(*output->outputType(), *exchange->outputType());
  EXPECT_EQ(*output->outputType()->childAt(0), *VARBINARY());
  EXPECT_EQ(*output->outputType()->childAt(2), *DECIMAL(38, 2));
  EXPECT_EQ(*producer->outputType()->childAt(2), *VARBINARY());
  EXPECT_EQ(*gpuConsumer->outputType(), *consumer->outputType());
  EXPECT_EQ(CudfPlanRewriter::rewrite(gpuProducer), gpuProducer);
  EXPECT_EQ(CudfPlanRewriter::rewrite(gpuConsumer), gpuConsumer);
  ASSERT_NE(
      dynamic_cast<const core::ProjectNode*>(gpuProducer->sources()[0].get()),
      nullptr);
  ASSERT_NE(
      dynamic_cast<const core::ProjectNode*>(gpuConsumer->sources()[0].get()),
      nullptr);
  EXPECT_EQ(
      *gpuConsumer->sources()[0]->outputType()->childAt(0), *DECIMAL(38, 2));

  Type::registerSerDe();
  core::ITypedExpr::registerSerDe();
  core::PlanNode::registerSerDe();
  CudfAggregationNode::registerSerDe();
  auto copy = ISerializable::deserialize<core::PlanNode>(
      gpuConsumer->serialize(), pool());
  auto* aggregation = dynamic_cast<const CudfAggregationNode*>(copy.get());
  ASSERT_NE(aggregation, nullptr);
  EXPECT_TRUE(aggregation->usesCompactDecimalSum(0));
  EXPECT_FALSE(aggregation->usesCompactDecimalSum(1));
  EXPECT_EQ(*aggregation->intermediateType(0), *DECIMAL(38, 2));
  EXPECT_EQ(copy->serialize(), gpuConsumer->serialize());
  auto producerCopy = ISerializable::deserialize<core::PlanNode>(
      gpuProducer->serialize(), pool());
  EXPECT_EQ(producerCopy->serialize(), gpuProducer->serialize());
  EXPECT_EQ(CudfPlanRewriter::rewrite(copy)->serialize(), copy->serialize());
  EXPECT_EQ(
      CudfPlanRewriter::rewrite(producerCopy)->serialize(),
      producerCopy->serialize());
}

TEST_F(CudfDecimalTest, compactSumRejectsGpuDisabledWithCpuFallback) {
  enableCpuFallback();
  auto input =
      makeRowVector({"d"}, {makeFlatVector<int64_t>({100}, DECIMAL(12, 2))});
  auto plan = exec::test::PlanBuilder()
                  .values({input})
                  .partialAggregation({}, {"sum(d) AS s"})
                  .finalAggregation()
                  .planNode();
  auto rewritten = CudfPlanRewriter::rewrite(plan);
  ASSERT_NE(dynamic_cast<const CudfAggregationNode*>(rewritten.get()), nullptr);
  VELOX_ASSERT_THROW(
      exec::test::AssertQueryBuilder(rewritten)
          .config(CudfConfig::kCudfEnabled, false)
          .copyResults(pool()),
      "Compact aggregation state requires cuDF to be enabled");
}

TEST_F(CudfDecimalTest, compactSumRejectsUnsupportedAggregateWithCpuFallback) {
  enableCpuFallback();
  auto input = makeRowVector(
      {"k", "d"},
      {makeFlatVector<int32_t>({1}),
       makeFlatVector<int64_t>({100}, DECIMAL(12, 2))});
  auto plan =
      exec::test::PlanBuilder()
          .values({input})
          .partialAggregation({"k"}, {"sum(d) AS s", "array_agg(k) AS a"})
          .finalAggregation()
          .planNode();
  auto rewritten = CudfPlanRewriter::rewrite(plan);
  auto* aggregation = dynamic_cast<const CudfAggregationNode*>(rewritten.get());
  ASSERT_NE(aggregation, nullptr);
  EXPECT_TRUE(aggregation->usesCompactDecimalSum(0));
  EXPECT_FALSE(aggregation->usesCompactDecimalSum(1));
  VELOX_ASSERT_THROW(
      exec::test::AssertQueryBuilder(rewritten)
          .config(CudfConfig::kCudfEnabled, true)
          .copyResults(pool()),
      "Compact aggregation state requires GPU execution; CPU fallback is not supported");
}

TEST_F(CudfDecimalTest, compactSumRejectsUnsupportedBoundariesAndModifiers) {
  auto input = makeRowVector(
      {"d", "m"},
      {makeFlatVector<int64_t>({100}, DECIMAL(12, 2)),
       makeFlatVector<bool>({true})});
  auto ids = std::make_shared<core::PlanNodeIdGenerator>();
  auto partial = exec::test::PlanBuilder(ids)
                     .values({input})
                     .partialAggregation({}, {"sum(d) AS s"})
                     .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(partial),
      "Compact aggregation state requires UCX partitioned output");
  auto output =
      exec::test::PlanBuilder(partial, ids).partitionedOutput({}, 1).planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(output),
      "Compact aggregation state cannot leave a task over non-UCX output");
  auto consumer = exec::test::PlanBuilder()
                      .exchange(partial->outputType(), "Presto")
                      .finalAggregation({}, {"sum(s) AS s"}, {{DECIMAL(12, 2)}})
                      .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(consumer),
      "Compact aggregation state requires an unordered UCX exchange");
  auto masked = exec::test::PlanBuilder()
                    .values({input})
                    .partialAggregation({}, {"sum(d) AS s"}, {"m"})
                    .finalAggregation()
                    .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(masked),
      "Compact Decimal64 SUM does not support DISTINCT, ordering or masks");
  auto projected =
      exec::test::PlanBuilder(partial, ids)
          .project({"coalesce(s, s) AS s"})
          .finalAggregation({}, {"sum(s) AS s"}, {{DECIMAL(12, 2)}})
          .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(projected),
      "Compact aggregation state requires a direct-field Project");

  auto projectedState =
      exec::test::PlanBuilder(partial, ids).project({"s AS state"}).planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(projectedState),
      "Compact aggregation state requires UCX partitioned output");
  auto projectedOutput = exec::test::PlanBuilder(projectedState, ids)
                             .partitionedOutput({}, 1)
                             .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(projectedOutput),
      "Compact aggregation state cannot leave a task over non-UCX output");
  auto droppedInput = exec::test::PlanBuilder()
                          .exchange(
                              ROW({"s", "unused"}, {VARBINARY(), VARBINARY()}),
                              "Presto",
                              std::string{core::TransportKind::kUcx})
                          .project({"s"})
                          .finalAggregation(
                              {}, {"sum(s) AS s"}, {{DECIMAL(12, 2)}})
                          .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(droppedInput),
      "Compact aggregation Project cannot drop input column: unused");
  auto conflictingAliases =
      exec::test::PlanBuilder(partial, ids)
          .project({"s AS sum_state", "s AS avg_state"})
          .finalAggregation(
              {},
              {"sum(sum_state) AS s", "avg(avg_state) AS a"},
              {{DECIMAL(12, 2)}, {DECIMAL(12, 2)}})
          .planNode();
  VELOX_ASSERT_THROW(
      CudfPlanRewriter::rewrite(conflictingAliases),
      "Conflicting compact state requirements for s");
}

TEST_F(CudfDecimalTest, reduceDecimal64SumCountSlicesAndNulls) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  const std::vector<int64_t> values = {999, 100, 999, -100, 999};
  const std::vector<bool> valid = {false, true, false, true, false};
  auto input = makeDecimalColumn<int64_t>(values, 2, &valid, stream);
  auto slices = cudf::slice(input->view(), {1, 4, 2, 3, 0, 0}, stream);
  const std::vector<int64_t> expectedCounts = {2, 0, 0};
  for (size_t i = 0; i < slices.size(); ++i) {
    SCOPED_TRACE(i);
    auto result = reduceDecimal64SumCount(slices[i], stream, mr);
    EXPECT_EQ(
        result.sum->type(), cudf::data_type(cudf::type_id::DECIMAL128, -2));
    EXPECT_EQ(copyColumnData<int128_t>(result.sum->view(), stream)[0], 0);
    EXPECT_EQ(
        copyColumnData<int64_t>(result.count->view(), stream)[0],
        expectedCounts[i]);
    EXPECT_EQ(result.sum->null_count(), expectedCounts[i] == 0 ? 1 : 0);
    EXPECT_EQ(result.count->null_count(), 0);
  }

  constexpr int64_t kBig = 900'000'000'000'000'000;
  auto nonNullInput = makeDecimalColumn<int64_t>(
      std::vector<int64_t>(12, kBig), 2, nullptr, stream);
  auto result = reduceDecimal64SumCount(nonNullInput->view(), stream, mr);
  EXPECT_EQ(
      copyColumnData<int128_t>(result.sum->view(), stream)[0],
      static_cast<int128_t>(kBig) * 12);
  EXPECT_EQ(copyColumnData<int64_t>(result.count->view(), stream)[0], 12);
  EXPECT_EQ(result.sum->null_count(), 0);
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateDecimal64) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<int64_t> sums = {100, -200, 300};
  std::vector<int64_t> counts = {1, 2, 0};
  std::vector<bool> sumValid = {true, false, true};
  std::vector<bool> countValid = {true, true, true};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);
  auto sumAndCount =
      deserializeDecimalSumState(stateCol->view(), 2, stream, get_temp_mr());
  auto stateMask = copyNullMask(stateCol->view(), stream);
  auto sumMask = copyNullMask(sumAndCount.sum->view(), stream);
  EXPECT_EQ(stateMask, sumMask);

  auto outSum = copyColumnData<__int128_t>(sumAndCount.sum->view(), stream);
  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(sumMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outSum[i], static_cast<__int128_t>(sums[i]));
    }
  }
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateDecimal128) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<__int128_t> sums = {
      static_cast<__int128_t>(123450),
      static_cast<__int128_t>(-25000),
      static_cast<__int128_t>(100000),
  };
  std::vector<int64_t> counts = {2, 1, 0};
  std::vector<bool> sumValid = {true, true, true};
  std::vector<bool> countValid = {true, false, true};

  auto sumCol = makeDecimalColumn<__int128_t>(sums, 3, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);
  auto sumAndCount =
      deserializeDecimalSumState(stateCol->view(), 3, stream, get_temp_mr());
  auto stateMask = copyNullMask(stateCol->view(), stream);
  auto sumMask = copyNullMask(sumAndCount.sum->view(), stream);
  EXPECT_EQ(stateMask, sumMask);

  auto outSum = copyColumnData<__int128_t>(sumAndCount.sum->view(), stream);
  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(sumMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outSum[i], sums[i]);
    }
  }
}

// Reproduces the Q18 failure scenario: serializes a decimal sum state with
// partial nulls, round-trips through Arrow (which compacts null rows to 0-byte
// payloads), then deserializes. Without the null-count fix in
// deserializeDecimalSumState, the payload size check would fire because
// chars_size == (numRows - nullCount) * 32, not numRows * 32.
TEST_F(CudfDecimalTest, decimalDeserializeSumStatePartialNullCompact) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  // Serialize a 3-row state where row 1 is null (sum-null, count == 0).
  std::vector<int64_t> sums = {100, 0, 300};
  std::vector<int64_t> counts = {1, 0, 2};
  std::vector<bool> sumValid = {true, false, true};
  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);

  // Round-trip through Arrow (cuDF -> Velox VARBINARY -> cuDF STRING).
  // Arrow stores null rows with 0-byte payloads, so the resulting cuDF STRING
  // column has chars_size == (numRows - nullCount) * 32.
  auto expectedType = ROW({{"s", VARBINARY()}});
  auto veloxRow = with_arrow::toVeloxColumn(
      cudf::table_view{{stateCol->view()}},
      pool(),
      expectedType,
      "s",
      stream,
      mr);
  auto compactTable = with_arrow::toCudfTable(veloxRow, pool(), stream, mr);
  auto compactStateView = compactTable->view().column(0);

  // Verify the column now has the compact layout.
  cudf::strings_column_view strings(compactStateView);
  EXPECT_LT(
      strings.chars_size(stream),
      static_cast<int64_t>(sums.size()) * 32); // 32 == kDecimalSumStateSize

  auto result =
      deserializeDecimalSumState(compactStateView, 2, stream, get_temp_mr());

  auto outSum = copyColumnData<__int128_t>(result.sum->view(), stream);
  auto outCount = copyColumnData<int64_t>(result.count->view(), stream);
  auto outMask = copyNullMask(result.sum->view(), stream);

  EXPECT_TRUE(isValidAt(outMask, 0));
  EXPECT_FALSE(isValidAt(outMask, 1));
  EXPECT_TRUE(isValidAt(outMask, 2));
  EXPECT_EQ(outSum[0], static_cast<__int128_t>(100));
  EXPECT_EQ(outCount[0], 1);
  EXPECT_EQ(outSum[2], static_cast<__int128_t>(300));
  EXPECT_EQ(outCount[2], 2);
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateUncompactedSlice) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  std::vector<int64_t> sums = {10, 20, 0, 40, 50};
  std::vector<int64_t> counts = {1, 2, 0, 4, 5};
  std::vector<bool> sumValid = {true, true, false, true, true};
  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);

  cudf::strings_column_view strings(stateCol->view());
  EXPECT_EQ(
      strings.chars_size(stream),
      static_cast<int64_t>(sums.size()) * 32); // 32 == kDecimalSumStateSize

  auto slices = cudf::slice(stateCol->view(), {1, 4});
  ASSERT_EQ(slices.size(), 1);
  ASSERT_EQ(slices.front().offset(), 1);

  auto result =
      deserializeDecimalSumState(slices.front(), 2, stream, get_temp_mr());
  auto outSum = copyColumnData<__int128_t>(result.sum->view(), stream);
  auto outCount = copyColumnData<int64_t>(result.count->view(), stream);
  auto outMask = copyNullMask(result.sum->view(), stream);

  ASSERT_EQ(outSum.size(), 3);
  EXPECT_TRUE(isValidAt(outMask, 0));
  EXPECT_FALSE(isValidAt(outMask, 1));
  EXPECT_TRUE(isValidAt(outMask, 2));
  EXPECT_EQ(outSum[0], static_cast<__int128_t>(20));
  EXPECT_EQ(outCount[0], 2);
  EXPECT_EQ(outSum[2], static_cast<__int128_t>(40));
  EXPECT_EQ(outCount[2], 4);
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateArrowCompactedSlice) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  // Slice the sum/count inputs first to pin serialization offset handling.
  // Round-trip through Arrow so the null row has a 0-byte payload, then slice
  // the state so deserialization must shift both offsets and validity bits.
  std::vector<int64_t> sums = {-999, 10, 20, 0, 40, 50, 999};
  std::vector<int64_t> counts = {9, 1, 2, 0, 4, 5, 9};
  std::vector<bool> sumValid = {true, true, true, false, true, true, true};
  auto sumParent = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countParent = makeInt64Column(counts, nullptr, stream);
  auto sumSlices = cudf::slice(sumParent->view(), {1, 6});
  auto countSlices = cudf::slice(countParent->view(), {1, 6});
  ASSERT_EQ(sumSlices.size(), 1);
  ASSERT_EQ(countSlices.size(), 1);
  ASSERT_EQ(sumSlices.front().offset(), 1);
  ASSERT_EQ(countSlices.front().offset(), 1);
  auto stateCol = serializeDecimalSumState(
      sumSlices.front(), countSlices.front(), stream, mr);

  auto expectedType = ROW({{"s", VARBINARY()}});
  auto veloxRow = with_arrow::toVeloxColumn(
      cudf::table_view{{stateCol->view()}},
      pool(),
      expectedType,
      "s",
      stream,
      mr);
  auto compactTable = with_arrow::toCudfTable(veloxRow, pool(), stream, mr);
  auto compactStateView = compactTable->view().column(0);
  cudf::strings_column_view strings(compactStateView);
  EXPECT_LT(
      strings.chars_size(stream), static_cast<int64_t>(stateCol->size()) * 32);

  auto slices = cudf::slice(compactStateView, {1, 4});
  ASSERT_EQ(slices.size(), 1);
  ASSERT_EQ(slices.front().offset(), 1);

  auto result =
      deserializeDecimalSumState(slices.front(), 2, stream, get_temp_mr());
  auto outSum = copyColumnData<__int128_t>(result.sum->view(), stream);
  auto outCount = copyColumnData<int64_t>(result.count->view(), stream);
  auto outMask = copyNullMask(result.sum->view(), stream);

  ASSERT_EQ(outSum.size(), 3);
  EXPECT_TRUE(isValidAt(outMask, 0));
  EXPECT_FALSE(isValidAt(outMask, 1));
  EXPECT_TRUE(isValidAt(outMask, 2));
  EXPECT_EQ(outSum[0], static_cast<__int128_t>(20));
  EXPECT_EQ(outCount[0], 2);
  EXPECT_EQ(outSum[2], static_cast<__int128_t>(40));
  EXPECT_EQ(outCount[2], 4);
}

// Reproduces the TPC-DS Q1 failure in CudfGroupbyFINAL: the streaming final
// aggregation concatenates its buffered result -- which came straight from
// serializeDecimalSumState and so keeps a 32-byte payload for null rows -- with
// each newly arrived batch, which was round tripped through velox and so has
// its null rows compacted to 0 bytes. The concatenated state column mixes both
// encodings, so its payload size is neither numRows * 32 nor
// (numRows - nullCount) * 32 and a two-way equality check rejects it.
TEST_F(CudfDecimalTest, decimalDeserializeSumStateMixedNullEncodings) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  auto serialize = [&](const std::vector<int64_t>& sums,
                       const std::vector<int64_t>& counts,
                       const std::vector<bool>& sumValid) {
    auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
    auto countCol = makeInt64Column(counts, nullptr, stream);
    return serializeDecimalSumState(
        sumCol->view(), countCol->view(), stream, mr);
  };

  // Buffered side: row 1 is null and keeps its 32-byte payload.
  auto fullState = serialize({100, 0, 300}, {1, 0, 2}, {true, false, true});

  // Incoming side: same shape, but the velox round trip compacts its null row
  // to a 0-byte payload.
  auto incomingState = serialize({400, 0, 600}, {3, 0, 4}, {true, false, true});
  auto veloxRow = with_arrow::toVeloxColumn(
      cudf::table_view{{incomingState->view()}},
      pool(),
      ROW({{"s", VARBINARY()}}),
      "s",
      stream,
      mr);
  auto compactTable = with_arrow::toCudfTable(veloxRow, pool(), stream, mr);

  auto mixed = cudf::concatenate(
      std::vector<cudf::column_view>{
          fullState->view(), compactTable->view().column(0)},
      stream,
      mr);

  // Four non-null rows, plus the one null payload the buffered side kept: the
  // payload size sits strictly between the compact and full extremes.
  cudf::strings_column_view mixedStrings(mixed->view());
  auto const numRows = static_cast<int64_t>(mixed->size());
  auto const nullCount = static_cast<int64_t>(mixed->null_count());
  EXPECT_EQ(numRows, 6);
  EXPECT_EQ(nullCount, 2);
  EXPECT_GT(
      mixedStrings.chars_size(stream),
      (numRows - nullCount) * 32); // 32 == kDecimalSumStateSize
  EXPECT_LT(mixedStrings.chars_size(stream), numRows * 32);

  auto result =
      deserializeDecimalSumState(mixed->view(), 2, stream, get_temp_mr());

  auto outSum = copyColumnData<__int128_t>(result.sum->view(), stream);
  auto outCount = copyColumnData<int64_t>(result.count->view(), stream);
  auto outMask = copyNullMask(result.sum->view(), stream);

  EXPECT_FALSE(isValidAt(outMask, 1));
  EXPECT_FALSE(isValidAt(outMask, 4));
  for (auto row : {0, 2, 3, 5}) {
    EXPECT_TRUE(isValidAt(outMask, row));
  }
  EXPECT_EQ(outSum[0], static_cast<__int128_t>(100));
  EXPECT_EQ(outCount[0], 1);
  EXPECT_EQ(outSum[2], static_cast<__int128_t>(300));
  EXPECT_EQ(outCount[2], 2);
  EXPECT_EQ(outSum[3], static_cast<__int128_t>(400));
  EXPECT_EQ(outCount[3], 3);
  EXPECT_EQ(outSum[5], static_cast<__int128_t>(600));
  EXPECT_EQ(outCount[5], 4);
}

// Trailing null: the offset for the last row equals chars_size, so the kernel
// would read 32 bytes past the buffer end without the null-mask guard.
TEST_F(CudfDecimalTest, decimalDeserializeSumStateTrailingNullCompact) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  // 3-row state: [valid, valid, null].
  std::vector<int64_t> sums = {100, 200, 0};
  std::vector<int64_t> counts = {1, 2, 0};
  std::vector<bool> sumValid = {true, true, false};
  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);

  auto expectedType = ROW({{"s", VARBINARY()}});
  auto veloxRow = with_arrow::toVeloxColumn(
      cudf::table_view{{stateCol->view()}},
      pool(),
      expectedType,
      "s",
      stream,
      mr);
  auto compactTable = with_arrow::toCudfTable(veloxRow, pool(), stream, mr);
  auto compactStateView = compactTable->view().column(0);

  cudf::strings_column_view strings(compactStateView);
  EXPECT_LT(strings.chars_size(stream), static_cast<int64_t>(sums.size()) * 32);

  auto result =
      deserializeDecimalSumState(compactStateView, 2, stream, get_temp_mr());

  auto outSum = copyColumnData<__int128_t>(result.sum->view(), stream);
  auto outCount = copyColumnData<int64_t>(result.count->view(), stream);
  auto outMask = copyNullMask(result.sum->view(), stream);

  EXPECT_TRUE(isValidAt(outMask, 0));
  EXPECT_TRUE(isValidAt(outMask, 1));
  EXPECT_FALSE(isValidAt(outMask, 2));
  EXPECT_EQ(outSum[0], static_cast<__int128_t>(100));
  EXPECT_EQ(outCount[0], 1);
  EXPECT_EQ(outSum[1], static_cast<__int128_t>(200));
  EXPECT_EQ(outCount[1], 2);
}

// Leading null: offset 0 overlaps the next valid row's data so the read is
// in-bounds, but verify the null mask propagates correctly.
TEST_F(CudfDecimalTest, decimalDeserializeSumStateLeadingNullCompact) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  // 3-row state: [null, valid, valid].
  std::vector<int64_t> sums = {0, 200, 300};
  std::vector<int64_t> counts = {0, 2, 3};
  std::vector<bool> sumValid = {false, true, true};
  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);

  auto expectedType = ROW({{"s", VARBINARY()}});
  auto veloxRow = with_arrow::toVeloxColumn(
      cudf::table_view{{stateCol->view()}},
      pool(),
      expectedType,
      "s",
      stream,
      mr);
  auto compactTable = with_arrow::toCudfTable(veloxRow, pool(), stream, mr);
  auto compactStateView = compactTable->view().column(0);

  cudf::strings_column_view strings(compactStateView);
  EXPECT_LT(strings.chars_size(stream), static_cast<int64_t>(sums.size()) * 32);

  auto result =
      deserializeDecimalSumState(compactStateView, 2, stream, get_temp_mr());

  auto outSum = copyColumnData<__int128_t>(result.sum->view(), stream);
  auto outCount = copyColumnData<int64_t>(result.count->view(), stream);
  auto outMask = copyNullMask(result.sum->view(), stream);

  EXPECT_FALSE(isValidAt(outMask, 0));
  EXPECT_TRUE(isValidAt(outMask, 1));
  EXPECT_TRUE(isValidAt(outMask, 2));
  EXPECT_EQ(outSum[1], static_cast<__int128_t>(200));
  EXPECT_EQ(outCount[1], 2);
  EXPECT_EQ(outSum[2], static_cast<__int128_t>(300));
  EXPECT_EQ(outCount[2], 3);
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateAllNull) {
  auto stream = cudf::get_default_stream();
  constexpr cudf::size_type numRows = 4;

  auto offsetsCol = cudf::make_fixed_width_column(
      cudf::data_type{cudf::type_id::INT32},
      numRows + 1,
      cudf::mask_state::UNALLOCATED,
      stream);
  auto* offsetsPtr = offsetsCol->mutable_view().data<int32_t>();
  auto status = cudaMemsetAsync(
      offsetsPtr,
      0,
      static_cast<size_t>(numRows + 1) * sizeof(int32_t),
      stream.value());
  VELOX_CHECK_EQ(0, static_cast<int>(status));
  stream.synchronize();

  std::vector<bool> valid(numRows, false);
  auto [nullMask, nullCount] = makeNullMask(valid, stream);
  rmm::device_buffer charsBuf(0, stream);
  auto stateCol = cudf::make_strings_column(
      numRows,
      std::move(offsetsCol),
      std::move(charsBuf),
      nullCount,
      std::move(nullMask));

  auto sumAndCount =
      deserializeDecimalSumState(stateCol->view(), 2, stream, get_temp_mr());
  auto outSumView = sumAndCount.sum->view();
  auto outCountView = sumAndCount.count->view();

  EXPECT_EQ(outSumView.size(), numRows);
  EXPECT_EQ(outCountView.size(), numRows);
  EXPECT_EQ(outSumView.null_count(), numRows);
  EXPECT_EQ(outCountView.null_count(), numRows);

  auto outSumMask = copyNullMask(outSumView, stream);
  auto outCountMask = copyNullMask(outCountView, stream);
  for (size_t i = 0; i < static_cast<size_t>(numRows); ++i) {
    EXPECT_FALSE(isValidAt(outSumMask, i));
    EXPECT_FALSE(isValidAt(outCountMask, i));
  }
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateRejectsInvalidRowWidth) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  // The total payload size is valid, but neither row contains a complete
  // serialized state. This exercises the per-row check rather than the
  // aggregate payload-size check.
  auto offsetsCol = makeFixedWidthColumn<int32_t>(
      cudf::data_type{cudf::type_id::INT32}, {0, 16, 64}, nullptr, stream);
  rmm::device_buffer charsBuf(64, stream);
  auto stateCol = cudf::make_strings_column(
      2,
      std::move(offsetsCol),
      std::move(charsBuf),
      0,
      cudf::create_null_mask(0, cudf::mask_state::UNALLOCATED, stream, mr));

  VELOX_ASSERT_THROW(
      deserializeDecimalSumState(stateCol->view(), 2, stream, get_temp_mr()),
      "Decimal sum state requires every non-null row to be 32 bytes");
}

TEST_F(CudfDecimalTest, decimalDeserializeSumStateRejectsMisalignedRow) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();

  auto sumCol = makeDecimalColumn<int64_t>({200}, 2, nullptr, stream);
  auto countCol = makeInt64Column({2}, nullptr, stream);
  auto alignedState =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);
  cudf::strings_column_view alignedStrings(alignedState->view());

  // The only non-null row has the correct width, but starts four bytes into
  // the chars buffer. The aggregate payload checks still accept the layout:
  // 32 non-null bytes in a 64-byte payload across three rows.
  auto offsetsCol = makeFixedWidthColumn<int32_t>(
      cudf::data_type{cudf::type_id::INT32}, {0, 4, 36, 64}, nullptr, stream);
  rmm::device_buffer charsBuf(64, stream);
  auto status = cudaMemcpyAsync(
      static_cast<uint8_t*>(charsBuf.data()) + 4,
      alignedStrings.chars_begin(stream),
      32, // kDecimalSumStateSize
      cudaMemcpyDeviceToDevice,
      stream.get());
  VELOX_CHECK_EQ(0, static_cast<int>(status));
  std::vector<bool> valid = {false, true, false};
  auto [nullMask, nullCount] = makeNullMask(valid, stream);
  auto stateCol = cudf::make_strings_column(
      3,
      std::move(offsetsCol),
      std::move(charsBuf),
      nullCount,
      std::move(nullMask));

  VELOX_ASSERT_THROW(
      deserializeDecimalSumState(stateCol->view(), 2, stream, get_temp_mr()),
      "Decimal sum state requires every non-null row to be 32 bytes and 8-byte aligned");
}

TEST_F(CudfDecimalTest, decimalSerializeSumStateUsesInt64OffsetsWhenEnabled) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  ScopedEnvVar enableLargeStrings("LIBCUDF_LARGE_STRINGS_ENABLED", "1");
  ScopedEnvVar threshold("LIBCUDF_LARGE_STRINGS_THRESHOLD", "1");

  std::vector<int64_t> sums = {100, -200};
  std::vector<int64_t> counts = {1, 1};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, nullptr, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);

  cudf::strings_column_view strings(stateCol->view());
  EXPECT_EQ(strings.offsets().type().id(), cudf::type_id::INT64);
}

TEST_F(CudfDecimalTest, decimalSumStateRoundTripUsesInt64Offsets) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  ScopedEnvVar enableLargeStrings("LIBCUDF_LARGE_STRINGS_ENABLED", "1");
  ScopedEnvVar threshold("LIBCUDF_LARGE_STRINGS_THRESHOLD", "1");

  std::vector<int64_t> sums = {100, -200, 300, 400};
  std::vector<int64_t> counts = {1, 0, 2, 3};
  std::vector<bool> sumValid = {true, true, false, true};
  std::vector<bool> countValid = {true, true, true, false};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);

  cudf::strings_column_view strings(stateCol->view());
  EXPECT_EQ(strings.offsets().type().id(), cudf::type_id::INT64);

  auto sumAndCount =
      deserializeDecimalSumState(stateCol->view(), 2, stream, get_temp_mr());
  auto outSumView = sumAndCount.sum->view();
  auto outCountView = sumAndCount.count->view();
  auto outSum = copyColumnData<__int128_t>(outSumView, stream);
  auto outCount = copyColumnData<int64_t>(outCountView, stream);
  auto outSumMask = copyNullMask(outSumView, stream);
  auto outCountMask = copyNullMask(outCountView, stream);

  EXPECT_EQ(outSumView.size(), sums.size());
  EXPECT_EQ(outCountView.size(), counts.size());
  EXPECT_EQ(outSumMask, outCountMask);

  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(outSumMask, i), expectedValid);
    EXPECT_EQ(isValidAt(outCountMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outSum[i], static_cast<__int128_t>(sums[i]));
      EXPECT_EQ(outCount[i], counts[i]);
    }
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal64) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<int64_t> sums = {100, 105, 250, -125};
  std::vector<int64_t> counts = {4, 2, 0, 2};
  std::vector<bool> sumValid = {true, true, true, true};
  std::vector<bool> countValid = {true, false, true, true};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<int64_t>(avgCol->view(), stream);

  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return static_cast<int64_t>(out);
  };

  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(avgMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i]));
    }
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal64Slice) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<int64_t> parentSums = {999, 100, 105, 250, -125, -999};
  std::vector<int64_t> parentCounts = {9, 4, 2, 0, 2, 9};
  std::vector<bool> parentSumValid = {true, true, true, true, true, true};
  std::vector<bool> parentCountValid = {true, true, false, true, true, true};

  auto sumParent =
      makeDecimalColumn<int64_t>(parentSums, 2, &parentSumValid, stream);
  auto countParent = makeInt64Column(parentCounts, &parentCountValid, stream);
  auto sumSlices = cudf::slice(sumParent->view(), {1, 5});
  auto countSlices = cudf::slice(countParent->view(), {1, 5});
  ASSERT_EQ(sumSlices.size(), 1);
  ASSERT_EQ(countSlices.size(), 1);
  ASSERT_EQ(sumSlices.front().offset(), 1);
  ASSERT_EQ(countSlices.front().offset(), 1);

  auto avgCol =
      computeDecimalAverage(sumSlices.front(), countSlices.front(), stream, mr);
  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<int64_t>(avgCol->view(), stream);

  std::vector<int64_t> sums = {100, 105, 250, -125};
  std::vector<int64_t> counts = {4, 2, 0, 2};
  std::vector<bool> countValid = {true, false, true, true};
  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return static_cast<int64_t>(out);
  };

  ASSERT_EQ(outAvg.size(), sums.size());
  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(avgMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i]));
    }
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal128) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<__int128_t> sums = {
      static_cast<__int128_t>(123450),
      static_cast<__int128_t>(-25000),
      static_cast<__int128_t>(100000),
  };
  std::vector<int64_t> counts = {3, 2, 0};
  std::vector<bool> sumValid = {true, true, true};
  std::vector<bool> countValid = {true, true, true};

  auto sumCol = makeDecimalColumn<__int128_t>(sums, 3, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<__int128_t>(avgCol->view(), stream);

  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return out;
  };

  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(avgMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i]));
    }
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal64MostNegativeSum) {
  // Negating INT64_MIN in the signed type is overflow UB; the magnitude and
  // sign must be handled in the unsigned domain. avg of one INT64_MIN is
  // itself.
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  constexpr int64_t kMin = std::numeric_limits<int64_t>::min();
  std::vector<int64_t> sums = {kMin};
  std::vector<int64_t> counts = {1};
  std::vector<bool> valid = {true};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 0, &valid, stream);
  auto countCol = makeInt64Column(counts, &valid, stream);
  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  auto outAvg = copyColumnData<int64_t>(avgCol->view(), stream);
  EXPECT_EQ(outAvg[0], kMin);
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal128MostNegativeSum) {
  // Same regression at the __int128 boundary. avg of one -2^127 is itself.
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  const __int128_t kMin =
      static_cast<__int128_t>(static_cast<unsigned __int128>(1) << 127);
  std::vector<__int128_t> sums = {kMin};
  std::vector<int64_t> counts = {1};
  std::vector<bool> valid = {true};

  auto sumCol = makeDecimalColumn<__int128_t>(sums, 0, &valid, stream);
  auto countCol = makeInt64Column(counts, &valid, stream);
  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  auto outAvg = copyColumnData<__int128_t>(avgCol->view(), stream);
  EXPECT_EQ(outAvg[0], kMin);
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal64AllValid) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<int64_t> sums = {100, 200, -150};
  std::vector<int64_t> counts = {4, 5, 3};
  std::vector<bool> sumValid = {true, true, true};
  std::vector<bool> countValid = {true, true, true};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  EXPECT_EQ(avgCol->view().null_count(), 0);
  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<int64_t>(avgCol->view(), stream);

  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return static_cast<int64_t>(out);
  };

  for (size_t i = 0; i < sums.size(); ++i) {
    if (avgCol->view().nullable()) {
      EXPECT_TRUE(isValidAt(avgMask, i)) << "row " << i;
    }
    EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i])) << "row " << i;
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal128AllValid) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<__int128_t> sums = {
      static_cast<__int128_t>(90000),
      static_cast<__int128_t>(-5000),
      static_cast<__int128_t>(1),
  };
  std::vector<int64_t> counts = {3, 5, 1};
  std::vector<bool> sumValid = {true, true, true};
  std::vector<bool> countValid = {true, true, true};

  auto sumCol = makeDecimalColumn<__int128_t>(sums, 3, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  EXPECT_EQ(avgCol->view().null_count(), 0);
  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<__int128_t>(avgCol->view(), stream);

  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return out;
  };

  for (size_t i = 0; i < sums.size(); ++i) {
    if (avgCol->view().nullable()) {
      EXPECT_TRUE(isValidAt(avgMask, i)) << "row " << i;
    }
    EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i])) << "row " << i;
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal64NonNullableInputs) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<int64_t> sums = {80, -40, 1000};
  std::vector<int64_t> counts = {2, 4, 10};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, nullptr, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  ASSERT_FALSE(sumCol->nullable());
  ASSERT_FALSE(countCol->nullable());

  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  EXPECT_EQ(avgCol->view().null_count(), 0);
  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<int64_t>(avgCol->view(), stream);

  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return static_cast<int64_t>(out);
  };

  for (size_t i = 0; i < sums.size(); ++i) {
    if (avgCol->view().nullable()) {
      EXPECT_TRUE(isValidAt(avgMask, i)) << "row " << i;
    }
    EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i])) << "row " << i;
  }
}

TEST_F(CudfDecimalTest, decimalComputeAverageDecimal128NonNullableInputs) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<__int128_t> sums = {
      static_cast<__int128_t>(600),
      static_cast<__int128_t>(-99),
  };
  std::vector<int64_t> counts = {3, 9};

  auto sumCol = makeDecimalColumn<__int128_t>(sums, 4, nullptr, stream);
  auto countCol = makeInt64Column(counts, nullptr, stream);
  ASSERT_FALSE(sumCol->nullable());
  ASSERT_FALSE(countCol->nullable());

  auto avgCol =
      computeDecimalAverage(sumCol->view(), countCol->view(), stream, mr);

  EXPECT_EQ(avgCol->view().null_count(), 0);
  auto avgMask = copyNullMask(avgCol->view(), stream);
  auto outAvg = copyColumnData<__int128_t>(avgCol->view(), stream);

  auto avgUnscaled = [](int128_t sum, int64_t count) {
    __int128_t out = 0;
    facebook::velox::DecimalUtil::
        divideWithRoundUp<__int128_t, __int128_t, int64_t>(
            out, sum, count, false, 0, 0);
    return out;
  };

  for (size_t i = 0; i < sums.size(); ++i) {
    if (avgCol->view().nullable()) {
      EXPECT_TRUE(isValidAt(avgMask, i)) << "row " << i;
    }
    EXPECT_EQ(outAvg[i], avgUnscaled(sums[i], counts[i])) << "row " << i;
  }
}

TEST_F(CudfDecimalTest, decimalSumStateRoundTripDecimal64) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<int64_t> sums = {100, -200, 300, 400};
  std::vector<int64_t> counts = {1, 0, 2, 3};
  std::vector<bool> sumValid = {true, true, false, true};
  std::vector<bool> countValid = {true, true, true, false};

  auto sumCol = makeDecimalColumn<int64_t>(sums, 2, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);
  auto stateMask = copyNullMask(stateCol->view(), stream);

  auto sumAndCount =
      deserializeDecimalSumState(stateCol->view(), 2, stream, get_temp_mr());
  auto outSumView = sumAndCount.sum->view();
  auto outCountView = sumAndCount.count->view();
  auto outSum = copyColumnData<__int128_t>(outSumView, stream);
  auto outCount = copyColumnData<int64_t>(outCountView, stream);
  auto outSumMask = copyNullMask(outSumView, stream);
  auto outCountMask = copyNullMask(outCountView, stream);

  EXPECT_EQ(stateMask, outSumMask);
  EXPECT_EQ(stateMask, outCountMask);

  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(stateMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outSum[i], static_cast<__int128_t>(sums[i]));
      EXPECT_EQ(outCount[i], counts[i]);
    }
  }
}

TEST_F(CudfDecimalTest, decimalSumStateRoundTripDecimal128) {
  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  std::vector<__int128_t> sums = {
      static_cast<__int128_t>(123450),
      static_cast<__int128_t>(-25000),
      static_cast<__int128_t>(100000),
  };
  std::vector<int64_t> counts = {2, 1, 0};
  std::vector<bool> sumValid = {true, false, true};
  std::vector<bool> countValid = {true, true, true};

  auto sumCol = makeDecimalColumn<__int128_t>(sums, 3, &sumValid, stream);
  auto countCol = makeInt64Column(counts, &countValid, stream);
  auto stateCol =
      serializeDecimalSumState(sumCol->view(), countCol->view(), stream, mr);
  auto stateMask = copyNullMask(stateCol->view(), stream);

  auto sumAndCount =
      deserializeDecimalSumState(stateCol->view(), 3, stream, get_temp_mr());
  auto outSumView = sumAndCount.sum->view();
  auto outCountView = sumAndCount.count->view();
  auto outSum = copyColumnData<__int128_t>(outSumView, stream);
  auto outCount = copyColumnData<int64_t>(outCountView, stream);
  auto outSumMask = copyNullMask(outSumView, stream);
  auto outCountMask = copyNullMask(outCountView, stream);

  EXPECT_EQ(stateMask, outSumMask);
  EXPECT_EQ(stateMask, outCountMask);

  for (size_t i = 0; i < sums.size(); ++i) {
    bool expectedValid = sumValid[i] && countValid[i] && counts[i] != 0;
    EXPECT_EQ(isValidAt(stateMask, i), expectedValid);
    if (expectedValid) {
      EXPECT_EQ(outSum[i], sums[i]);
      EXPECT_EQ(outCount[i], counts[i]);
    }
  }
}

TEST_F(CudfDecimalTest, cudfVarbinaryArrowRoundTrip) {
  auto input = makeRowVector(
      {"bin"},
      {makeNullableFlatVector<std::string>(
          {std::string("abc"), std::nullopt, std::string("xyz")},
          VARBINARY())});

  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  auto cudfTable = with_arrow::toCudfTable(input, pool(), stream, mr);
  auto roundTrip =
      with_arrow::toVeloxColumn(cudfTable->view(), pool(), "rt_", stream, mr);

  ASSERT_EQ(roundTrip->childAt(0)->type()->kind(), TypeKind::VARCHAR);
  VELOX_ASSERT_THROW(
      roundTrip->setType(ROW({{"rt_0", VARBINARY()}})),
      "Cannot change vector type");

  auto expected = makeRowVector(
      {"rt_0"},
      {makeNullableFlatVector<std::string>(
          {std::string("abc"), std::nullopt, std::string("xyz")}, VARCHAR())});

  facebook::velox::test::assertEqualVectors(expected, roundTrip);
}

TEST_F(CudfDecimalTest, cudfVarbinaryArrowRoundTripWithExpectedType) {
  auto input = makeRowVector(
      {"bin"},
      {makeNullableFlatVector<std::string>(
          {std::string("abc"), std::nullopt, std::string("xyz")},
          VARBINARY())});

  auto expectedType = ROW({{"bin", VARBINARY()}});

  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  auto cudfTable = with_arrow::toCudfTable(input, pool(), stream, mr);
  auto roundTrip = with_arrow::toVeloxColumn(
      cudfTable->view(), pool(), expectedType, "rt_", stream, mr);

  ASSERT_EQ(roundTrip->childAt(0)->type()->kind(), TypeKind::VARBINARY);

  auto expected = makeRowVector(
      {"rt_0"},
      {makeNullableFlatVector<std::string>(
          {std::string("abc"), std::nullopt, std::string("xyz")},
          VARBINARY())});

  facebook::velox::test::assertEqualVectors(expected, roundTrip);
}

TEST_F(CudfDecimalTest, cudfVarbinaryRowTypeMismatch) {
  auto input = makeRowVector(
      {"l_returnflag",
       "l_linestatus",
       "avg_51",
       "avg_52",
       "avg_53",
       "count_54",
       "sum_47",
       "sum_48",
       "sum_49",
       "sum_50"},
      {makeFlatVector<std::string>({"A", "B"}, VARCHAR()),
       makeFlatVector<std::string>({"F", "O"}, VARCHAR()),
       makeFlatVector<std::string>({"x", "y"}, VARBINARY()),
       makeFlatVector<std::string>({"p", "q"}, VARBINARY()),
       makeFlatVector<std::string>({"m", "n"}, VARBINARY()),
       makeFlatVector<int64_t>({10, 20}, BIGINT()),
       makeFlatVector<std::string>({"u", "v"}, VARBINARY()),
       makeFlatVector<std::string>({"r", "s"}, VARBINARY()),
       makeFlatVector<std::string>({"t", "w"}, VARBINARY()),
       makeFlatVector<std::string>({"c", "d"}, VARBINARY())});

  auto expectedType = ROW({
      {"l_returnflag", VARCHAR()},
      {"l_linestatus", VARCHAR()},
      {"avg_51", VARBINARY()},
      {"avg_52", VARBINARY()},
      {"avg_53", VARBINARY()},
      {"count_54", BIGINT()},
      {"sum_47", VARBINARY()},
      {"sum_48", VARBINARY()},
      {"sum_49", VARBINARY()},
      {"sum_50", VARBINARY()},
  });

  auto stream = cudf::get_default_stream();
  auto mr = cudf::get_current_device_resource_ref();
  auto cudfTable = with_arrow::toCudfTable(input, pool(), stream, mr);
  auto roundTrip =
      with_arrow::toVeloxColumn(cudfTable->view(), pool(), "rt_", stream, mr);

  ASSERT_EQ(roundTrip->childAt(2)->type()->kind(), TypeKind::VARCHAR);
  ASSERT_EQ(roundTrip->childAt(3)->type()->kind(), TypeKind::VARCHAR);
  ASSERT_EQ(roundTrip->childAt(4)->type()->kind(), TypeKind::VARCHAR);
  ASSERT_EQ(roundTrip->childAt(6)->type()->kind(), TypeKind::VARCHAR);
  ASSERT_EQ(roundTrip->childAt(7)->type()->kind(), TypeKind::VARCHAR);
  ASSERT_EQ(roundTrip->childAt(8)->type()->kind(), TypeKind::VARCHAR);
  ASSERT_EQ(roundTrip->childAt(9)->type()->kind(), TypeKind::VARCHAR);

  VELOX_ASSERT_THROW(
      roundTrip->setType(expectedType), "Cannot change vector type");
}

} // namespace
} // namespace facebook::velox::cudf_velox
